lib/elixir/lib/code/normalizer.ex

```diff
@@ -0,0 +1,246 @@
+defmodule Code.Normalizer do
+  @moduledoc false
+
+  defguard is_literal(x)
+           when is_integer(x) or
+                  is_float(x) or
+                  is_binary(x) or
+                  is_atom(x)
+
+  @doc """
+  Wraps literals in the quoted expression to conform to the AST format expected
+  by the formatter.
+  """
+  def normalize(quoted, opts \\ []) do
+    line = Keyword.get(opts, :line, 1)
+
+    do_normalize(quoted, line: line)
+  end
+
+  # Skip already wrapped blocks
+  defp do_normalize({:__block__, _, [literal]} = quoted, _parent_meta) when is_literal(literal) do
+    quoted
+  end
+
+  defp do_normalize({:__block__, meta, args} = quoted, _parent_meta) do
+    if Keyword.has_key?(meta, :format) do
+      quoted
+    else
+      {:__block__, meta, Enum.map(args, &do_normalize(&1, meta))}
+    end
+  end
+
+  # Skip aliases so the module segment atoms don't get wrapped
+  defp do_normalize({:__aliases__, _, _} = quoted, _parent_meta) do
+    quoted
+  end
+
+  # Skip qualified tuples left hand side
+  defp do_normalize({:., _, [_, :{}]} = quoted, _parent_meta) do
+    quoted
+  end
+
+  # Don't normalize the `Access` atom in access syntax
+  defp do_normalize({:., _, [Access, :get]} = quoted, _parent_meta) do
+    quoted
+  end
+
+  # Only normalize the left side of the dot operator
+  # The right hand side is an atom in the AST but it's not an atom literal, so
+  # it should not be wrapped
+  defp do_normalize({:., meta, [left, right]}, _parent_meta) do
+    {:., meta, [do_normalize(left, meta), right]}
+  end
+
+  # A list of left to right arrows is not considered as a list literal, so it's
+  # not wrapped
+  defp do_normalize([{:->, _, _} | _] = quoted, parent_meta) do
+    Enum.map(quoted, &do_normalize(&1, parent_meta))
+  end
+
+  # left -> right
+  defp do_normalize({:->, meta, [left, right]}, _parent_meta) do
+    left = Enum.map(left, &do_normalize(&1, meta))
+    right = do_normalize(right, meta)
+    {:->, meta, [left, right]}
+  end
+
+  # Maps
+  defp do_normalize({:%{}, meta, args}, _parent_meta) do
+    args = Enum.map(args, &do_normalize(&1, meta))
+
+    args =
+      case args do
+        # Unwrap the right hand side if we're in an update syntax
+        [{:|, pipe_meta, [left, {_, _, [right]}]}] ->
+          [{:|, pipe_meta, [left, right]}]
+
+        # Unwrap args so we have 2-tuples instead of blocks
+        [{_, _, args}] ->
+          args
+
+        args ->
+          args
+      end
+
+    {:%{}, meta, args}
+  end
+
+  # If a keyword list is an argument of a guard, we need to drop the block
+  # wrapping
+  defp do_normalize({:when, meta, args} = _quoted, _parent_meta) do
+    args =
+      Enum.map(args, fn
+        arg when is_list(arg) ->
+          {_, _, [arg]} = do_normalize(arg, meta)
```

Comment:
Maybe break normalize in two? `do_normalize` handles literals and then calls `do_normalize_non_literal`, which you would also invoke here?

Reply:
Another option is to make things consistent and to change the parser and make sure the args argument is also wrapped in the literal block when it is a keyword, and change the formatter to handle it accordingly. I think I might prefer that. Thoughts?

Reply:
I think consistency for those cases would be nice. I realized the problem in this case is more general than `a when b: :c`. If the keyword list has square brackets, then the arguments get wrapped in a block like a regular list, but if the brackets are omitted then the arguments don't get wrapped:
```elixir
# a when b: :c
{:when, [line: 1],
 [
   {:a, [line: 1], nil},
   [
     {{:__block__, [format: :keyword, line: 1], [:b]},
      {:__block__, [line: 1], [:c]}}
   ]
 ]}

# a when [b: :c]
{:when, [line: 1],
 [
   {:a, [line: 1], nil},
   {:__block__, [closing: [line: 1], line: 1],
    [
      [
        {{:__block__, [format: :keyword, line: 1], [:b]},
         {:__block__, [line: 1], [:c]}}
      ]
    ]}
 ]}

# foo(a, b: :c) do :d end
{:foo,
 [
   do: [line: 1],
   end: [line: 1],
   closing: [line: 1],
   line: 1 
 ],
 [
   {:a, [line: 1], nil},
   [
     {{:__block__, [format: :keyword, line: 1], [:b]},
      {:__block__, [line: 1], [:c]}}
   ],
   [
     {{:__block__, [line: 1], [:do]},
      {:__block__, [line: 1], [:d]}}
   ]
 ]}

# foo(a, [b: :c]) do :d end
{:foo,
 [
   do: [line: 1],
   end: [line: 1],
   closing: [line: 1],
   line: 1 
 ],
 [
   {:a, [line: 1], nil},
   {:__block__, [closing: [line: 1], line: 1],
    [
      [
        {{:__block__, [format: :keyword, line: 1], [:b]},
         {:__block__, [line: 1], [:c]}}
      ]
    ]},
   [
     {{:__block__, [line: 1], [:do]}, 
      {:__block__, [line: 1], [:d]}}
   ]
 ]}
```

I did the unwrapping in some places because it depends on the context(a standalone keyword list needs the wrapping but it's not needed in cases like the ones mentioned above). If the parser did that work instead, this could be avoided altogether.
But how can the formatter know the user preference in this case? With the current behavior it is trivial to know if the user omitted the square brackets or not, just check if it's wrapped in a block, but if everything is wrapped in a block there would be no way to tell the difference?

Reply:
Maybe the normalizer can just wrap everything in a block and not do any unwrapping, since it lacks the context to know if the brackets should be omitted or not anyways?

Reply:
The formatter will remove the wrapping brackets when it can, so it should be fine to always wrap it. We will know for sure once we add the more integration like tests. :)